tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -12,90 +12,457 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  // Potential signature mismatch - generate conversions if needed.
+  TargetFunctionInfo(const string callee_name_, bool check_signature_,
+                     absl::Span<const PrimitiveType> input_types_,
+                     const PrimitiveType output_type_)
+      : callee_name(callee_name_),
+        check_signature(check_signature_),
+        input_types(input_types_),
+        result_type(output_type_) {}
+  // Potential signature mistamatch - use bitcast if specified by the target.
+  TargetFunctionInfo(const string callee_name_, bool check_signature_,
+                     absl::Span<const PrimitiveType> input_types_,
+                     const PrimitiveType output_type_, bool use_bitcast_)
+      : callee_name(callee_name_),
+        check_signature(check_signature_),
+        input_types(input_types_),
+        result_type(output_type_),
+        use_bitcast(use_bitcast_) {}
+  // Device function name.
+  const string callee_name;
+  // Check signature of the device function and if needed generate casts.
+  bool check_signature;
+  // Inpute types accespted by the device function.
+  absl::Span<const PrimitiveType> input_types;
+  // Result type of the device function.
+  PrimitiveType result_type;
+  // Use bitcast to generate casts if the desired signature at the call site
+  // does not match the signature of the device function.
+  absl::optional<bool> use_bitcast;
+};
+
+// Wrapper structure to carry either information about the intrinsic
+// or device function for NVPTX/AMDGPU.
+// Note: Only intrinsic or device function - not both - is expected.
+struct TargetInfo {
+  absl::optional<struct TargetIntrinsicInfo> target_intrinsic_info;
+  absl::optional<struct TargetFunctionInfo> target_function_info;
+  TargetInfo(struct TargetIntrinsicInfo x) : target_intrinsic_info(x) {}
+  TargetInfo(struct TargetFunctionInfo y) : target_function_info(y) {}
 };
 
-// Gets the llvm intrinsic ids on different platforms (NVPTX, AMDGPU)
-// corresponding to the give TargetIntrinsicID.
-struct TargetIntrinsics GetIntrinsic(TargetIntrinsicID intrin) {
-  switch (intrin) {
-    case TargetIntrinsicID::kShflDownF32: {
-      return {llvm::Intrinsic::nvvm_shfl_sync_down_f32,
-              llvm::Intrinsic::not_intrinsic};
+// Populates the function information for different platforms (NVPTX, AMDGPU)
+// corresponding to the given TargetFunctionID.
+// The call site provides arguments that is least restrictive to both targets.
+// For instance, if one target provides device function that takes 2 arguments
+// and the other provides arguments that takes 4 arguments, then the call
+// site provides four arguments. The following function  takes care of ignoring
+// the two superfluous argumenrs for the GPU device function that takes 2
+// arguments. In a similar vein, the following function also specifies
+// if conversions need to be generated for any argument type mismatches.
+// See check_signature  and use_bitcast fields in TargetFunctionInfo
```

Comment:
I do not understand the second paragraph (starting at "For instance").

Do you have someone on your side who can proofread the comments (and the code, while we're at it)?  It will reduce a lot of back-and-forth.

Reply:
> Hi, thank you for the PR!

Thanks for the comments! 

> Can you point me at the code that will use this?

The target util provides a way to abstract the calls to either target intrinsics or device functions. While in this PR, I have includes all the changed needed in target_util  for all the device functions or intrinsic  like the math calls, shfl functions and the various intrinsics, only code from ir_emission and intrinsic calls  use them. In a subsequent PR, I plan to submit changes in elemental_ir_emitter that will also use this functionality from target_util. 

If you would prefer, I can limit this PR to only those functions/intrinsics that are used and skip the math functionality for now.  I just wanted to show the complete extent of functionality desired.

Reply:
> I do not understand the second paragraph (starting at "For instance").

For shfl, rocm  uses a device function that takes two arguments while ptx uses an intrinsic that takes 4 arguments. So while the call site provides a way to provide for 4 arguments, target_util needs to provide a way for ignoring the extra arguments for rocm. I used PRIMITIVE_TYPE_INVALID to represent those arguments.

Reply:
> Can you point me at the code that will use this?

Part of the reason I ask is because I'm trying to understand whether the math changes are even necessary.  The NVPTX LLVM backend knows how to lower llvm math intrinsics, and surely the AMDGCN backend could know how to do this if it doesn't already.

>> I do not understand the second paragraph (starting at "For instance").

Thank you for the explanation.

FYI in general when I say "I don't understand X" what I'm looking for is a change to the comments or code to help me and future readers understand it.  Explanations in the PR are less useful because they disappear once people stop looking here.

For this particular bit I don't think silently ignoring two arguments is a good API.  It is unintuitive, hard to use, and hard to read.  It's also not terribly flexible -- what if we had a case where the two APIs took e.g. the same number of arguments but in a different order, or if the args meant different things?

To me, it seems that TargetUtil should be used for cases where there are NVPTX and AMDGCN functions that are basically equivalent.  In this case they're not, because the NVPTX version takes additional args.

Instead, could we have an "emit shfl down" function which emits the two-arg version on amdgpu and the four-arg version on nvptx?  This function might use TargetUtil, or it might not, because it's already `switch`'ing on the target anyway.

Reply:
> Part of the reason I ask is because I'm trying to understand whether the math changes are even necessary. The NVPTX LLVM backend knows how to lower llvm math intrinsics, and surely the AMDGCN backend could know how to do this if it doesn't already.

My guess is both targets prefer device functions because its easier to provide for different implementations to satisfy tricky  requirements such as ulp etc instead of lowerings inside the backend. 


> To me, it seems that TargetUtil should be used for cases where there are NVPTX and AMDGCN functions that are basically equivalent. In this case they're not, because the NVPTX version takes additional args.
> Instead, could we have an "emit shfl down" function which emits the two-arg version on amdgpu and the four-arg version on nvptx? This function might use TargetUtil, or it might not, because it's already `switch`'ing on the target anyway.

Sure...so I guess then there is no need for providing the signature etc then. The idea would be to just call different implementations when there is a mismatch.